src/Disks/ObjectStorages/MetadataStorageFromPlainObjectStorageOperations.cpp

```diff
@@ -549,4 +549,105 @@ void MetadataStorageFromPlainObjectStorageMoveFileOperation::finalize()
+
+    std::unordered_set<std::string> subdirs = {""};
+    path_map.iterateSubdirectories(from.string() + "/", [&](const auto & elem){ subdirs.emplace(elem); });
+    for (const auto & subdir : subdirs)
+    {
+        auto sub_path_to = to / subdir;
+        auto sub_path_from = from / subdir;
+
+        path_map.print();
+        LOG_TRACE(getLogger("MetadataStorageFromPlainObjectStorageRemoveRecursiveOperation"), "Moving '{}' to '{}'", sub_path_from, sub_path_to);
+
+        /// parent_path() removes the trailing '/'.
+        path_map.moveDirectory(sub_path_from.parent_path(), sub_path_to.parent_path());
+    }
+}
+
+void MetadataStorageFromPlainObjectStorageRemoveRecursiveOperation::execute()
+{
+    if (path_map.existsLocalPath(path))
+    {
+        inMemoryTreeMove(path, tmp_path);
+        moved = true;
+    }
+}
+
+void MetadataStorageFromPlainObjectStorageRemoveRecursiveOperation::undo()
+{
+    if (moved)
+    {
+        inMemoryTreeMove(tmp_path, path);
+    }
+}
+
+void MetadataStorageFromPlainObjectStorageRemoveRecursiveOperation::finalize()
+{
+    if (!moved)
+        return;
+
+    StoredObjects objects_to_remove;
+
+    std::unordered_set<std::string> subdirs = {""};
+    path_map.iterateSubdirectories(tmp_path.string() + "/", [&](const auto & elem){ subdirs.emplace(elem); });
+    for (const auto & subdir : subdirs)
+    {
+        auto subdir_path = tmp_path / subdir;
+        LOG_TRACE(getLogger("MetadataStorageFromPlainObjectStorageRemoveRecursiveOperation"), "Removing directory '{}'", subdir_path);
+
+        /// Info should exist since it's lifetime is bounded to execution of this operation, because tmp path is unique.
+        path_map.print();
+        auto directory_info = path_map.getRemotePathInfoIfExists(subdir_path.parent_path()).value();
+        auto metadata_object_key = createMetadataObjectKey(directory_info.path, metadata_key_prefix);
+        objects_to_remove.emplace_back(metadata_object_key.serialize(), path / PREFIX_PATH_FILE_NAME);
+
+        /// We also need to remove all files inside each of the subdirectories.
+        std::unordered_set<std::string> files = {};
+        path_map.iterateFiles(subdir_path, [&](const auto & elem){ files.emplace(elem); });
+        for (const auto & file : files)
+        {
+            auto file_path = subdir_path / file;
+            LOG_TRACE(getLogger("MetadataStorageFromPlainObjectStorageRemoveRecursiveOperation"), "Removing file '{}'", file_path);
+
+            auto file_object_key = object_storage->generateObjectKeyForPath(file_path, std::nullopt).serialize();
+            objects_to_remove.emplace_back(file_object_key, file_path);
+        }
+    }
+
+    /// Clear in-memory state of removed tree
+    for (const auto & subdir : subdirs)
+    {
+        auto subdir_path = tmp_path / subdir;
+        path_map.print();
+        bool is_removed = path_map.removePathIfExists(subdir_path.parent_path());
```

Comment:
I guess this will work even if you iterate from parent to child directories?

Reply:
Yes, it should be, because the path map does not validate the tree structure.